File: src/main/java/org/micoli/php/ui/panels/ListRowFilter.java
package org.micoli.php.ui.panels;

import com.intellij.openapi.diagnostic.Logger;
import java.util.Arrays;
import java.util.List;
import java.util.regex.Pattern;
import javax.swing.*;
import org.micoli.php.symfony.list.SearchableRecord;

public class ListRowFilter<M, I> extends RowFilter<M, I> {
    protected static final Logger LOGGER = Logger.getInstance(ListRowFilter.class);
    private String searchText = "";
    private List<Pattern> searchPatterns = null;
    private List<String> searchParts = null;
    private boolean isRegexMode = true;

    public void updateFilter(String searchText, boolean isRegexMode) {
        this.searchText = searchText;
        this.isRegexMode = isRegexMode;
        searchPatterns = isRegexMode
                ? Arrays.stream(searchText.split(" "))
                        .map(String::trim)
                        .map(expr -> Pattern.compile("(?i)" + expr))
                        .toList()
                : null;
        searchParts = !isRegexMode
                ? Arrays.stream(searchText.split(" ")).map(String::trim).toList()
                : null;
    }

    @Override
    public boolean include(Entry<? extends M, ? extends I> entry) {
        if (searchText.isEmpty()) {
            return true;
        }
        if (entry.getValue(entry.getValueCount() - 1) instanceof SearchableRecord searchableRecord) {
            if (isRegexMode) {
                return searchPatterns.stream().allMatch(pattern -> pattern.matcher(searchableRecord.getSearchString())
                        .find());
            }
            return searchParts.stream()
                    .allMatch(part ->
                            searchableRecord.getSearchString().toLowerCase().contains(part.toLowerCase()));
        }
        return false;
    }
}

File: src/main/kotlin/org/micoli/php/ui/panels/ListRowFilter.kt
package org.micoli.php.ui.panels

import javax.swing.RowFilter
import kotlin.Boolean
import kotlin.text.isEmpty
import org.micoli.php.symfony.list.SearchableRecord
import org.micoli.php.ui.panels.rowMatchers.EmptyMatcher
import org.micoli.php.ui.panels.rowMatchers.PlainMatcher
import org.micoli.php.ui.panels.rowMatchers.RegexMatcher
import org.micoli.php.ui.panels.rowMatchers.RowMatcher

class ListRowFilter<M, I> : RowFilter<M?, I?>() {
    var searcher: RowMatcher = EmptyMatcher()

    fun updateFilter(searchText: String, isRegexMode: Boolean) {
        searcher =
            when {
                searchText.isEmpty() -> EmptyMatcher()
                isRegexMode -> RegexMatcher(searchText)
                !isRegexMode -> PlainMatcher(searchText)
                else -> EmptyMatcher()
            }
    }

    override fun include(entry: Entry<out M?, out I?>): Boolean {
        val value = entry.getValue(entry.valueCount - 1)
        if (value !is SearchableRecord) {
            return false
        }
        return searcher.match(value)
    }
}
